pageserver/src/layered_repository/metadata.rs

```diff
@@ -18,18 +19,23 @@ use zenith_utils::{
 use crate::config::PageServerConf;
 
 // Taken from PG_CONTROL_MAX_SAFE_SIZE
-const METADATA_MAX_SAFE_SIZE: usize = 512;
-const METADATA_CHECKSUM_SIZE: usize = std::mem::size_of::<u32>();
-const METADATA_MAX_DATA_SIZE: usize = METADATA_MAX_SAFE_SIZE - METADATA_CHECKSUM_SIZE;
-
+const METADATA_MAX_SIZE: usize = 512;
+const METADATA_HDR_SIZE: usize = std::mem::size_of::<TimelineMetadataHeader>();
```

Comment:
https://doc.rust-lang.org/std/mem/fn.size_of.html

says that

> In general, the size of a type is not stable across compilations, but specific types such as primitives are.

Might be a bit dangerous, `#[repr(C)]` can be used for better compatibility guarantees.
No danger with the current structure contents though.

Reply:
But we do not need some mapping on C struct here. So even if rust compiler will allocate this struct in some `strange` way (different fields order or unexpected alignment), then still this code will work.

Reply:
> So even if rust compiler will allocate this struct in some strange way (different fields order or unexpected alignment), then still this code will work.

I'm talking about the _**size**_ though, not the alignment or field order. The latter does not matter much (if the size does not change) since we're dealing with bytes through serde.

But the fact that `TimelineMetadataHeader::des(&metadata_bytes[0..METADATA_HDR_SIZE])?` could theoretically break from a regular rustc update (while `format_version` would be the same) is now subtle and could be dangerous in the future.

-----

Right now the header contains primitives only though and it's fine, since primitives and various `Option`, `Rc`, etc. are guaranteed to have the same size, as the doc says, so we have to put something else inside the header first to make its serialization "unstable". Afaik, adding any collection such as `Vec` is enough already for that though.

So, not insisting on any change, but feel like `repr(C)` (and a comment, optionally?) might be trivial enough to add to ensure that nothing idd happens a few years later here.

Reply:
Struct size mostly depends on fields order and alignment.
But even if rust compiler does something strange (actually it doesn't) and size of this structure will be more than 8 bytes, than nothing bad will happen: there is no assumption  in the code that size should be exactly 8 bytes.

Yes, if compiler/serde behavior will be changed in further releases and so format of structure is changed, then we will get the problem. But probability of it is IMHO very very small. And it is not the only place when we are using serde and so depend on its behavior. I do not understand what should we write in comment: "serialization format does matter"? But it seems to be obvious without comments... Actually almost in every place where serialization is used, serialized bytes are stored in some persistent media and so changing format or serialization rules will cause incompatibility. We can handle it through storage format version (and that is the goal of this patch). We should just not to forget to increment STORAGE_FORMAT_VERSION when something is changed.